### PR TITLE
add package ipaddress

### PR DESCRIPTION
**Pull request by Luffbee**
# ipaddress
https://github.com/VladimirShaleev/ipaddress/tree/main

A library for working and manipulating IPv4/IPv6 addresses and networks in modern C++.

This cross-platfrom header-only library (for C++11 and newer) is inspired by the [ipaddress API in Python](https://docs.python.org/3.12/library/ipaddress.html), from which it derives its name. It aims to be simpler to use due to its familiar interface. However, the C++ implementation takes a different approach: it uses static polymorphism through the strategy pattern instead of dynamic polymorphism to handle differences between IP versions (IPv4 and IPv6). This design choice eliminates the overhead of dynamic calls and virtual tables. For instance, an instance of the ipv4_address class will be represented by 4 bytes.

The library leverages modern C++ features, ensuring that all IP address and network operations support constant expressions.

### CODE REVIEW

**Comment by SirLynix** [comment]
`ipaddress` is a very generic name, the package should probably be called `vladimirshaleev_ipaddress` (similar to what vcpkg does)

File: packages/i/ipaddress/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    add_urls("https://github.com/VladimirShaleev/ipaddress/archive/refs/tags/$(version).tar.gz",
+        "https://github.com/VladimirShaleev/ipaddress.git")
+
+    add_versions("v1.1.0", "e5084d83ebd712210882eb6dac14ed1b9b71584dede523b35c6181e0a06375f1")
+
+    add_configs("noexcept", {
+        description = "Disable handling cpp exception for",
+        default = false,
+        type = "boolean"
+    })
```
Comment (waruqi) [inline packages/i/ipaddress/xmake.lua:18]: use single line for each config

File: packages/i/ipaddress/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        table.insert(configs, "-DIPADDRESS_IPV6_SCOPE_MAX_LENGTH=" .. package:config("ipv6_scope_max_length"))
+
+        import("package.tools.cmake").install(package, configs)
+    end)
+
+    on_test(function(package)
+        assert(package:check_cxxsnippets({
+            test = [[
+            #include <iostream>
+
```
Comment (waruqi) [inline packages/i/ipaddress/xmake.lua:57]: remove empty line

File: packages/i/ipaddress/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            void subnets_sample() {
+                constexpr auto net = ipv4_network::parse("192.0.2.0/24");
+
+                std::cout << "subnets for " << net << ':' << std::endl;
+                for (const auto& subnet : net.subnets(2)) {
+                    std::cout << "  " << subnet << std::endl;
+                }
+
+                constexpr auto last_subnet = net.subnets(2).back();
```
Comment (waruqi) [inline packages/i/ipaddress/xmake.lua:85]: too large test code, please simplify it

File: packages/i/ipaddress/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,95 @@
+package("ipaddress", function()
+    set_homepage("https://vladimirshaleev.github.io/ipaddress/")
+    set_description("A library for working and manipulating IPv4/IPv6 addresses and networks")
+    set_license("MIT")
+    set_kind("library", {
```
Comment (waruqi) [inline packages/i/ipaddress/xmake.lua:5]: ues single line

File: packages/i/ipaddress/xmake.lua (hunk 1 above)
Comment (Luffbee) [reply packages/i/ipaddress/xmake.lua:18]: Is there any format tool for xmake? This file is formatted by the Lua extension os vscode.

Comment (waruqi) [reply packages/i/ipaddress/xmake.lua:18]: no tool, you can see other packages

Comment (waruqi) [reply packages/i/ipaddress/xmake.lua:18]: please do not any format tool

File: packages/i/ipaddress/xmake.lua (hunk 3 above)
Comment (Luffbee) [reply packages/i/ipaddress/xmake.lua]: how about now?

**Comment by Luffbee** [comment]
> `ipaddress` is a very generic name, the package should probably be called `vladimirshaleev_ipaddress` (similar to what vcpkg does)

renamed to the same name as vcpkg

File: packages/v/vladimirshaleev-ipaddress/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    set_license("MIT")
+    set_kind("library", {headeronly = true})
+
+    add_urls("https://github.com/VladimirShaleev/ipaddress/archive/refs/tags/$(version).tar.gz",
+        "https://github.com/VladimirShaleev/ipaddress.git")
+
+    add_versions("v1.1.0", "e5084d83ebd712210882eb6dac14ed1b9b71584dede523b35c6181e0a06375f1")
+
+    add_configs("noexcept", {description = "Disable handling cpp exception for", default = false, type = "boolean"})
+    add_configs("no_overload_std", {description = "Do not overload std functions such as to_string, hash etc", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/v/vladimirshaleev-ipaddress/xmake.lua:13]: remove `no_` prefix, default -> true

File: packages/v/vladimirshaleev-ipaddress/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    add_configs("no_ipv6_scope", {description = "Disable scope id for IPv6 addresses", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/v/vladimirshaleev-ipaddress/xmake.lua:14]: and here

File: packages/v/vladimirshaleev-ipaddress/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    set_description("A library for working and manipulating IPv4/IPv6 addresses and networks")
+    set_license("MIT")
+    set_kind("library", {headeronly = true})
+
+    add_urls("https://github.com/VladimirShaleev/ipaddress/archive/refs/tags/$(version).tar.gz",
+        "https://github.com/VladimirShaleev/ipaddress.git")
+
+    add_versions("v1.1.0", "e5084d83ebd712210882eb6dac14ed1b9b71584dede523b35c6181e0a06375f1")
+
+    add_configs("noexcept", {description = "Disable handling cpp exception for", default = false, type = "boolean"})
```
Comment (waruqi) [inline packages/v/vladimirshaleev-ipaddress/xmake.lua:12]: noexcept -> except

File: packages/v/vladimirshaleev-ipaddress/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            void teredo_sample() {
+                constexpr auto teredo_ip = "2001:0000:4136:e378:8000:63bf:3fff:fdd2"_ipv6;
+                auto [server, client] = teredo_ip.teredo().value();
+                std::cout << "server: " << server << " and client: " << client << " for " << teredo_ip << std::endl << std::endl;
+            }
+        ]]
+        }, {
+            configs = {
+                languages = "c++17"
+            }
```
Comment (waruqi) [inline packages/v/vladimirshaleev-ipaddress/xmake.lua:54]: single line config

File: packages/v/vladimirshaleev-ipaddress/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,56 @@
+package("vladimirshaleev-ipaddress")
+    set_homepage("https://vladimirshaleev.github.io/ipaddress/")
+    set_description("A library for working and manipulating IPv4/IPv6 addresses and networks")
+    set_license("MIT")
+    set_kind("library", {headeronly = true})
```
Comment (waruqi) [inline packages/v/vladimirshaleev-ipaddress/xmake.lua:5]: move it to line 2

File: packages/v/vladimirshaleev-ipaddress/xmake.lua (hunk 7 above)
Comment (Luffbee) [reply packages/v/vladimirshaleev-ipaddress/xmake.lua:12]: I searched similar configs, spdlog use 'noexcept' and some others use 'exceptions'. So maybe 'exceptions' is better?

**Comment by SirLynix** [comment]
I don't get it, if it's a header-only package, options are meaningless if they don't add any define to the package

**Comment by Luffbee** [comment]
> > I don't get it, if it's a header-only package, options are meaningless if they don't add any define to the package
> 
> you are right, they should be exported by `package:add("defines", "xxx")`
> 
> @Luffbee
> 
> https://github.com/VladimirShaleev/ipaddress/blob/fffbd6cb64df8be25de2e3d992771a0fa57bd97c/CMakeLists.txt#L29

Sorry, I thought xmake would extract these defines from cmake. I will fix it in another PR.

P.S. Is there any document for these contribution FAQs? (e.g. format, option naming, options of headeronly package). I did some searching but couldn't find any.

**Pull request merged**